File: foundations-jdbc-kotlin/src/kotlin/dev/typr/foundationskt/DbType.kt
@file:Suppress("unused")
package dev.typr.foundationskt

open class DbType<T>(open val underlying: dev.typr.foundations.DbType<T>) {
    open fun opt(): DbType<T?> =
        DbType(underlying.opt().to(Bijection.optionalToNullable()))

    open fun <B> to(bijection: dev.typr.foundations.Bijection<T, B>): DbType<B> =
        DbType(underlying.to(bijection))

    /** Whether this type allows null values. True for types created with [opt]. */
    fun isNullable(): Boolean = underlying.isNullable

    operator fun invoke(value: T): Fragment =
        Fragment(underlying.apply(value))

    override fun toString(): String = underlying.toString()
}

File: foundations-jdbc-scala/src/scala/dev/typr/foundationssc/DbType.scala
package dev.typr.foundationssc

open class DbType[T](val underlying: dev.typr.foundations.DbType[T]):
  def opt: DbType[Option[T]] =
    DbType(underlying.opt().to(Bijections.optionalToOption))

  def to[B](bijection: Bijection[T, B]): DbType[B] =
    DbType(underlying.to(bijection))

  /** Bind a value to this type, producing a Fragment suitable for `sql"..."` interpolation. */
  def apply(value: T): Fragment = Fragment.encode(this, value)

  /** Whether this type allows null values. True for types created with [[opt]]. */
  def isNullable: Boolean = underlying.isNullable

  override def toString: String = underlying.toString
